### ISSUE TITLE
LambdaMetaFactory generating invalid jimple bodies for constructor references

### ISSUE THREAD

**Issue by dseekatz** [soot-oss/soot#1367]
Hello!

This issue is possibly related to #1292.

I am using Soot 4.1.0 and openjdk 1.8.0_222, and I am trying to perform some analysis on Java code that uses constructor references:

```
public class A {

    public Supplier<Object> constructorReference() {
        return Object::new;
    }

}
```

Soot recognizes the resulting `invokedynamic` instruction and creates a Supplier instance via LambdaMetaFactory. The following jimple body is created for the supplier `get()` method:

```
    public java.lang.Object get()
    {
        A$init__1 $r0;
        java.lang.Object $r1;

        $r0 := @this: A$init__1;

        $r1 = new java.lang.Object;

        specialinvoke $r1.<java.lang.Object: void <init>()>();

        return;
    }
```

I would expect that the method should return $r1 (the newly created object), as per the method's signature, but a void return statement is inserted instead.  The [comments in LambdaMetaFactory](https://github.com/Sable/soot/blob/3b5bcb09f536e2676e9f781405f852f0f7cde249/src/main/java/soot/LambdaMetaFactory.java#L783-L785) indicate that the void return statement may be intentional in the specific case of a constructor reference.  This seems strange, since the method signature and return type do not match.

I've noticed that the Jimple validators allow for this sort of disagreement between method signature and return type, so I imagine there must be some cases where it's desirable.  For my use case, I am performing some analysis on the jimple representation and then generating transformed class files, so it is causing some issues for me with invalid bytecode.

Any idea if this is intentional behaviour or whether it's something I could fix?

**Comment by dseekatz** [comment]
I can add a simple check to `ReturnStatementsValidator` for void vs. non-void methods.  

@mbenz89 I'll also take a closer look at the if condition.  It seems to me that the return type of the functional interface and the return type of the implementing method should always both be void or both be non-void, _except_ in the case of a constructor reference.  I'll work on this today and hopefully it will be a straightforward fix.

**Comment by mbenz89** [comment]
fixed in #1368
